### PR TITLE
Normal estimation for closed surfaces

### PR DESCRIPTION
**Pull request by RainSureZhao**
Hello, I have successfully implemented the algorithm according to the algorithm steps you gave. After several experiments, the experimental results are good.

image

### CODE REVIEW

File: src/point_cloud/normal_estimator.cpp [hunk 1]
```diff
@@ -1,7 +1,170 @@
...
+  plane_estimator est(points_(nn_indices, Eigen::all));
 
+  if (est.plane_factor() < plane_factor_threshold) {
+    return geometry::vector3d::Zero();
+  }
+
+  return est.plane_normal();
+}
+
+}  // namespace polatory::point_cloud
```
Comment (unageek) [inline src/point_cloud/normal_estimator.cpp:162]: Hey, you are duplicating the code!

File: examples/estimates_normals/parse_options.hpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      ("p_outside_x", po::value(&opts.p_outside_x)->required()->value_name("VAL"),
+      "The X VAL of p_outside")
+      ("p_outside_y", po::value(&opts.p_outside_y)->required()->value_name("VAL"),
+       "The Y VAL of p_outside")
+      ("p_outside_z", po::value(&opts.p_outside_z)->required()->value_name("VAL"), 
+      "The Z VAL of p_outside")(
+      "min-plane-factor", po::value(&opts.min_plane_factor)->default_value(1.8)->value_name("VAL"),
+      "Threshold of acceptance for estimated normal vectors")
+      ("out", po::value(&opts.out_file)->required()->value_name("FILE"),
+      "Output file in CSV format:\n  X,Y,Z,VAL");
```
Comment (unageek) [inline examples/estimates_normals/parse_options.hpp:39]: `X,Y,Z,NX,NY,NZ`

**Review by unageek — CHANGES_REQUESTED** [review]
Hi! Thank you for sending the PR! Please check my comments.

**Comment by RainSureZhao** [comment]
> Hi! Thank you for sending the PR! Please check my comments.

Ok, copy that.

File: src/point_cloud/normal_estimator.cpp (hunk 1 above)
Comment (RainSureZhao) [reply src/point_cloud/normal_estimator.cpp:162]: > Hey, you are duplicating the code!

Sorry, due to a compiler error at run time after submitting PR, I had to try to copy the locally written code to a clone of a brand new repository to commit.🤐

File: src/point_cloud/normal_estimator.cpp [hunk 3]
```diff
@@ -42,6 +42,91 @@ normal_estimator& normal_estimator::estimate_with_radius(double radius,
...
+    point3d t = {P[0] - p_outside[0], P[1] - p_outside[1], P[2] - p_outside[2]};
+    if (n.dot(t) > 0.0) {
+      n = -n;
+    }
+
+    // 5. Propagate the orientation of normals using the nearest neighbor search
+    for (auto x : component) {
+      if (x == closest_index) continue;
+      auto m = normals_.row(x);
+      if (m.dot(n) < 0) {
```
[the remark this answers is not in the document]
Comment (RainSureZhao) [reply src/point_cloud/normal_estimator.cpp:285]: Hello, I have written the code according to the pseudocode implementation you provided, but the time complexity of this new method seems to be very high; When I tested the data, it was almost impossible to get results.

This is the code link: 
https://paste.ubuntu.com/p/yYTWx4fVHH/

Each step is annotated.

[the remark this answers is not in the document]
Comment (unageek) [reply src/point_cloud/normal_estimator.cpp:285]: Oops, same points are added to the queue multiple times. Try modifying the code as follows:

```
 1. Estimate (unoriented) normals for all points
 2. P ← all points (P tracks the unoriented points, use std::unordered_set<index_t>)
 3. while P ≠ ∅
 4.     p_closest ← closest point in P to p_outside
 5.     Align p_closest.normal with the vector p_outside - p_closest
 6.     Remove p_closest from P
 7.     Q ← {p_closest}
 8.     while Q ≠ ∅
 9.         p ← pop(Q)
10.         N ← kNN of p
11.         for q in N ∩ P
12.             Align q.normal with p.normal
13.             Remove q from P
14.             Q ← Q ∪ {q}
```

A few comments:

1. In this code:

   ```cpp
           auto q_normal = normals_.row(q);
           auto p_normal = normals_.row(p);
           // 12. Align q.normal with p.normal
           if (q_normal.dot(p_normal) < 0) {
             q_normal = -q_normal;
           }
   ```

   [Avoid](https://eigen.tuxfamily.org/dox/TopicPitfalls.html) assigning Eigen's expressions to auto variables. `normals_.row(p)` should be assigned to a `geometry::vector3d` and the expression `normals_.row(q)` should be inlined.

2. Distance computation is as simple as `(b - a).norm()`.

[the remark this answers is not in the document]
Comment (RainSureZhao) [reply src/point_cloud/normal_estimator.cpp:285]: Roger that! Thank you very much.

**Comment by RainSureZhao** [comment]
@unageek Hi, I checked the error message in the build, but it looked very strange, I did not see these errors in the uploaded code.

**Comment by unageek** [comment]
These errors are due to the code duplication.

**Comment by RainSureZhao** [comment]
> These errors are due to the code duplication.

Sorry, because I haven't used Github for a long time, so I can't use it well. So how do I solve this problem?

**Comment by unageek** [comment]
Please resolve the duplication and push the code again. 😄

**Pull request closed without merging**